pkg/result/processors/sort_results.go

```diff
@@ -0,0 +1,179 @@
+package processors
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/golangci/golangci-lint/pkg/config"
+	"github.com/golangci/golangci-lint/pkg/result"
+)
+
+// Base propose of this functionality to sort results (issues)
+// produced by various linters by analyzing code. We achieving this
+// by sorting results.Issues using processor step, and chain based
+// rules that can compare different properties of the Issues struct.
+
+var _ Processor = (*SortResults)(nil)
+
+type SortResults struct {
+	cmp Comparator
+	cfg *config.Config
+}
+
+func NewSortResults(cfg *config.Config) *SortResults {
+	// For sorting we are comparing (in next order): file names, line numbers,
+	// position, range numbers (if position not available), and finally - giving up.
+	return &SortResults{
+		cmp: ByName{
+			next: ByLine{
+				next: ByColumn{},
+			},
+		},
+		cfg: cfg,
+	}
+}
+
+// Process is performing sorting of the result issues.
+func (sr SortResults) Process(issues []result.Issue) ([]result.Issue, error) {
+	if !sr.cfg.Output.SortResults {
+		return issues, nil
+	}
+
+	sort.Slice(issues, func(i, j int) bool {
+		return sr.cmp.Compare(&issues[i], &issues[j]) == Less
+	})
+
+	return issues, nil
+}
+
+func (sr SortResults) Name() string { return "sort_results" }
+func (sr SortResults) Finish()      {}
+
+type CompareResult int
+
+const (
+	Less CompareResult = iota - 1
+	Equal
+	Greater
+	None
+)
+
+func (c CompareResult) isNeutral() bool {
+	// return true if compare result is incomparable or equal.
+	return c == None || c == Equal
+}
+
+func (c CompareResult) String() string {
+	switch c {
+	case Less:
+		return "Less"
+	case Equal:
+		return "Equal"
+	case Greater:
+		return "Greater"
+	}
+
+	return "None"
+}
+
+// Comparator describe how to implement compare for two "issues" lexicographically
+type Comparator interface {
+	Compare(a, b *result.Issue) CompareResult
+	Next() Comparator
+}
+
+var (
```

Comment:
Why is an empty interface instance created as nil but thrown away for each type? I know you can use this method to always ensure interface casting will work but I'm not familiar with doing this without ever storing the value. I removed these lines and line 16 and the tests still works just fine.

Is there any planned conversion going on or used elsewhere outside this PR? Even if so, isn't the "comma-ok" notation preferred (`byName, ok := someInterace.(*ByName)` and check for `ok`)?

Reply:
```go
var (
	_ Comparator = (*ByName)(nil)
	_ Comparator = (*ByLine)(nil)
	_ Comparator = (*ByColumn)(nil)
	_ Comparator = (*ByRange)(nil)
)
```

Commenting code on github sometimes can be confusing =). This is nil allocations interface checksum that will block compilation, you can find (almost) same looking code in almost every preprocessor, the only difference  `var _ Processor = SkipFiles{}` will escape to the heap.